# Математические пакеты.

## Лабораторная работа №3: Использование библиотеки символьных вычислений SymPy для решения задач

Выполнил студент группы 9381 Судаков Евгений, вариант 16.

## Цель работы

Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

1. Используя методы библиотеки SymPy решить диофантово уравнение;
2. Используя методы библиотеки SymPy по открытой части ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово;
3. Используя методы библиотеки SymPy записать в ДНФ и КНФ булеву функцию, заданную аналитическим выражением.

## Выполнение работы

Подключаем необходимы библиотеки:

In [8]:
import sympy as sp
import sympy.crypto.crypto as spc
import sympy.logic.boolalg as spl

### 1. Решение диофантового уравнения

Решим уравнение $2295x+ 4862y = −119$ с помощью метода sp.diophantine:

In [9]:
x, y = sp.symbols("x y")
sp.diophantine(2295 * x + 4862 * y + 119)

{(286*t_0 - 875, 413 - 135*t_0)}

Результат:

$x = 286 t_0 - 875$

$y = 413 - 135 t_0,$
 
$t_0 \in \mathbb{Z}$

### 2. Расшифровка закодированного сообщения

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 15,m = 46).
Шифрованное сообщение имеет вид (29; 4; 16; 36). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “e” и “ъ”).


In [21]:
key = spc.rsa_private_key(46, 15)
word = [29, 4, 16, 36]
letters = ['а', 'б', 'в', 'г', 'д', 'е', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н', 'о', 'п', 'р', 'с', 'т', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'щ', 'ы', 'ь', 'э', 'ю', 'я']

result = ''
for i in word:
    result += letters[spc.decipher_rsa(i, key) - 2]
print(result)

зрак


### 3. Преобразование булевой функции в ДНФ И КНФ

$ f(x, y, z) = ((yz) ∨ (z  \oplus  x))(x ∨ (x \oplus y)) $

In [25]:
from sympy.abc import x, y, z

f = ((y & z) | (z ^ x))&(x | (x ^ y))

print('ДНФ: ', spl.to_dnf(f, True))
print('КНФ: ', spl.to_cnf(f, True))

ДНФ:  (y & z) | (x & ~z)
КНФ:  (x | z) & (y | ~z)


### Выводы
Были получены навыки работы с пакетом SymPy для решения задач дискретной математики и математической логики, а именно:
1. Было решено диафонтово уравнение
2. С помощью алгоритма RSA декодировано зашифрованное сообщение
3. Булева функция была приведена к ДНФ и КНФ
